### ISSUE TITLE
Can I use it with ADAM?

### ISSUE THREAD

**Issue by lazysquid** [titu1994/Snapshot-Ensembles#8]
Hi. I have some questions before I'm going to use snapshot ensembles
TL;DR
1. Can I use snapshot-ensembles with adaptive learning rate optimizer? (e.g. ADAM)
2. Can I use it with very slow optimizer such as SGD?
3. Which optimizer and parameters shows best for your experience?
4. If I find optimal weight for my partial problem by using this technique, can I use it as initial point to find optimal point for full problem?
5. Does Snapshot-Ensembles affect to training time?

Situation:
I'm training a network which is based on FCN Densenet. (the problem is similar to image segmentation but continuous labels). Since I have quite large dataset(over 350GB), it is almost impossible to find proper parameter and model with this setting(almost 6 hour to train one epoch). Instead using full data, I am training it with small part of data(around 8~9GB). 

The problem is if I use ADAM, whenever I train model with either full or partial data, it converges really fast. It converges within 1 ~ 2 epoch when I train it with full data and it also converges within 10 ~ 30 epoch when I use small data. 
I'm guessing ADAM just put model into closest local optima. 
If I initialize weight differently at every epoch, maybe I can find several local optima with ADAM in really short time and I might have performance boost by using this technique.

But the problem is I don't know I can use it with ADAM. Is it possible?

**Comment by titu1994** [comment]
Are you sure you are converting within 1-2 epochs with Adam?

In any case, this technique should work with Adam as well, since it will just scale Adams final gradient computations. However, the improvements won't be as dramatic as using simple SGD. However, small improvements will still be there due to ensemble effect. 

If you are converting that fast, then you can set number of epochs to 10 or 20 instead of 300, and get a snapshot every 2-4 epochs. 

I personally have stopped using this technique since the relative performance improvement is not worth the linear increase in evaluation time. 

Since DenseNet FCN has an already large inference time, I would not suggest using this technique for production systems. For research, it is fine. 

To answer your questions in the order you stated above :
1) Yes, it can be used with Adam
2) Ofcourse. It was originally meant to help SGD in the first place. 
3) I've only tried on Cifar 10. There, the default settings of the paper worked the best. 
4) Yes you can initialize the model with pretrained weights. The improvements won't be much however even after ensembled predictions. 
5) It does not affect training time. However, it linearly increases prediction time if you use ensemble averaging over all snapshots (by a factor of how many snapshots there are)

**Comment by titu1994** [comment]
Do note that for point 4, it can cause a bad initialization for the full dataset training when the distribution of the full dataset is different from the distribution of the smaller dataset. 

To partially avoid this, try to sample your smaller dataset in the same proportions as the large dataset. But with semantic Segmentation, and such large quantities of data, I think the effect of poor initialization for the full set is going to be very small.
